### PR TITLE
Rollover max docs should only count primaries

### LINKED ISSUE #24217: rollover `max_docs` uses doc count from both primary and replica/s
**Elasticsearch version**: 5.3.0

The rollover `max_docs` value uses the doc count from both primary and replica/s.  It should be based upon the primary only, since the replica count can be changed.

(start two nodes, to allow the default 1 replica to allocate)
```
PUT /logs-000001 
{
  "aliases": {
    "logs_write": {}
  }
}
PUT logs_write/test/1
{
  "blah": "foo"
}
POST /logs_write/_rollover?dry_run
{
  "conditions": {
    "max_docs":  2
  }
}
# response
{
  "old_index": "logs-000001",
  "new_index": "logs-000002",
  "rolled_over": false,
  "dry_run": true,
  "acknowledged": false,
  "shards_acknowledged": false,
  "conditions": {
    "[max_docs: 2]": true
  }
}
```
change the replica count to 0:
```
PUT /logs-000001/_settings
{
  "index": {
    "number_of_replicas": 0
  }
}
# response
{
  "old_index": "logs-000001",
  "new_index": "logs-000002",
  "rolled_over": false,
  "dry_run": true,
  "acknowledged": false,
  "shards_acknowledged": false,
  "conditions": {
    "[max_docs: 2]": false
  }
}
```

### PR DESCRIPTION
**Pull request by fred84**
max_doc condition for index rollover should use document count only from primary shards 

Fixes #24217

### CODE REVIEW

**Comment by karmi** [comment]
Hi @fred84, we have found your signature in our records, but it seems like you have signed with a different e-mail than the one used in yout Git [commit](https://github.com/elastic/elasticsearch/pull/24977.patch). Can you please add both of these e-mails into your Github profile (they can be hidden), so we can match your e-mails to your Github profile?

File: rest-api-spec/src/main/resources/rest-api-spec/test/indices.rollover/20_max_doc_condition.yml [hunk 1]
```diff
@@ -0,0 +1,80 @@
+---
+"Rollover conditions matched with replica node":
+  # create index with alias and replica
+  - do:
+      indices.create:
+        index: logs-1
+        wait_for_active_shards: all
```
Comment (bleskes) [inline rest-api-spec/src/main/resources/rest-api-spec/test/indices.rollover/20_max_doc_condition.yml]: I would much prefer it if you add a unit test in TransportRolloverActionTests instead. The REST yaml tests are there to make sure we pass requests and read responses correctly but it has a big overhead to test simple inner behavior with it. To do so you can add an overload of evaluateConditions that takes IndicesStatsResponse and translates it to a call to an evaluateConditions which gets the number of docs (it seems that's the only stats we use in [here](https://github.com/elastic/elasticsearch/blob/master/core/src/main/java/org/elasticsearch/action/admin/indices/rollover/TransportRolloverAction.java#L197)

**Review by bleskes — CHANGES_REQUESTED** [review]
Thank you for the PR. The change looks good but I would love to move the test to be a unit test. I left a suggestion on how to do it.

File: rest-api-spec/src/main/resources/rest-api-spec/test/indices.rollover/20_max_doc_condition.yml (hunk 1 above)
Comment (fred84) [inline rest-api-spec/src/main/resources/rest-api-spec/test/indices.rollover/20_max_doc_condition.yml]: Thanks for suggestion. I'll create unit test. Should I keep yaml test?

**Comment by fred84** [comment]
@karmi I have used incorrect email in commit. Should I create new PR?

File: rest-api-spec/src/main/resources/rest-api-spec/test/indices.rollover/20_max_doc_condition.yml (hunk 1 above)
Comment (s1monw) [inline rest-api-spec/src/main/resources/rest-api-spec/test/indices.rollover/20_max_doc_condition.yml]: yeah please keep it

**Comment by s1monw** [comment]
> @karmi I have used incorrect email in commit. Should I create new PR?

@fred84 you can change you setting for git and then rebase the PR and force push to you branch that should fix it?

File: rest-api-spec/src/main/resources/rest-api-spec/test/indices.rollover/20_max_doc_condition.yml (hunk 1 above)
Comment (fred84) [inline rest-api-spec/src/main/resources/rest-api-spec/test/indices.rollover/20_max_doc_condition.yml]: @bleskes I added 2 more tests to TransportRolloverActionTests

**Review by bleskes — CHANGES_REQUESTED** [review]
Thank you @fred84! I think we are moving in the right direction. I would like to suggest another simplifications. We already have a fairly extended test to test the condition logic (`testEvaluateConditions`). We don't really need to duplicate the testing logic in it. I think we need to test here is only the conversion between `IndicesStatsResponse` and the `DocStats` used for the conditions. We can do so, using the new method you added, by adding a single test call `testDocStatsSelection` (or something like) that has a single custom condition when asserts that the numbers you got are what you expect. You can then randomize the stats in a `IndicesStatsResponse` and use the custom condition to assert that the right ones (i.e., the primaries) have been passed into it.

**Comment by fred84** [comment]
Thanks for suggestions, @bleskes. I updated the PR. The only thing I disagree that we need randomize stats in test. "testDocStatsSelectionFromPrimariesOnly" tests only that right value passed from IndicesStatResponse to Condition. There is no manipulation with this value inside "evaluateConditions".

**Comment by s1monw** [comment]
@elasticmachine ok to test

**Comment by karmi** [comment]
@fred84, sorry, I've missed your notification, I can see the CLA check is green now, @s1monw's advice was spot on, thanks for updating the e-mail!

**Comment by s1monw** [comment]
I see this in the build logs:

```
{p0=indices.rollover/20_max_doc_condition/Rollover conditions matched with replica node}]: after test
  1> [2017-06-07T03:35:53,692][INFO ][o.e.t.r.IntegTestZipClientYamlTestSuiteIT] Stash dump on failure [{
  1>   "stash" : {
  1>     "body" : null
  1>   }
  1> }]
ERROR   30.0s | IntegTestZipClientYamlTestSuiteIT.test {p0=indices.rollover/20_max_doc_condition/Rollover conditions matched with replica node} <<< FAILURES!
   > Throwable #1: java.lang.RuntimeException: Failure at [indices.rollover/20_max_doc_condition:4]: listener timeout after waiting for [30000] ms
   > 	at __randomizedtesting.SeedInfo.seed([856786C5C842552:800247B6F27848AA]:0)
   > 	at org.elasticsearch.test.rest.yaml.ESClientYamlSuiteTestCase.executeSection(ESClientYamlSuiteTestCase.java:346)
   > 	at org.elasticsearch.test.rest.yaml.ESClientYamlSuiteTestCase.test(ESClientYamlSuiteTestCase.java:328)
   > 	at java.lang.Thread.run(Thread.java:748)
   > Caused by: java.io.IOException: listener timeout after waiting for [30000] ms
   > 	at org.elasticsearch.client.RestClient$SyncResponseListener.get(RestClient.java:660)
   > 	at org.elasticsearch.client.RestClient.performRequest(RestClient.java:219)
   > 	at org.elasticsearch.client.RestClient.performRequest(RestClient.java:191)
   > 	at org.elasticsearch.test.rest.yaml.ClientYamlTestClient.callApi(ClientYamlTestClient.java:169)
   > 	at org.elasticsearch.test.rest.yaml.ClientYamlTestExecutionContext.callApiInternal(ClientYamlTestExecutionContext.java:157)
   > 	at org.elasticsearch.test.rest.yaml.ClientYamlTestExecutionContext.callApi(ClientYamlTestExecutionContext.java:89)
   > 	at org.elasticsearch.test.rest.yaml.section.DoSection.execute(DoSection.java:221)
   > 	at org.elasticsearch.test.rest.yaml.ESClientYamlSuiteTestCase.executeSection(ESClientYamlSuiteTestCase.java:344)
   > 	... 37 more```

@fred84 can you take a look at it?

**Comment by bleskes** [comment]
The rest test are run (sometimes) against a 1 node cluster. This means that replicas won't always be assigned. This `wait_for_active_shards: all` condition in the rest test makes it fail. I suggest just removing it. We also probably don't need explicit setting replicas to 1. Better leave to the default imo.

**Comment by fred84** [comment]
@bleskes @s1monw This yml test make no sense when running without replica (and will fail anyway with different number of shards/replicas). Is it possible to run it on specific number of shards/replicas? Or better remove this test?

**Comment by s1monw** [comment]
> @bleskes @s1monw This yml test make no sense when running without replica (and will fail anyway with different number of shards/replicas). Is it possible to run it on specific number of shards/replicas? Or better remove this test?

we run test with one node and with multiple nodes I think your test will be fine it should just use defaults and remove the `wait_for_active_shards: all` condition to make sure it also passes in a single node enviroment

**Comment by bleskes** [comment]
> This yml test make no sense when running without replica

I agree it will be great if we can pin down the test to always fail without your fix. Sadly this is not possible today. Doing it like we suggested means it will fail *sometimes* (if the replicas were fast enough to allocate). That's better than nothing.

> (and will fail anyway with different number of shards/replicas)

I'm not sure I follow this one. can you clarify?

**Comment by fred84** [comment]
@bleskes 
>
> > (and will fail anyway with different number of shards/replicas)
>
> I'm not sure I follow this one. can you clarify?

Test had checks for docs count in both "total" and "primaries": 
```yml 
- match:    { _all.primaries.docs.count: 1 }
- match:    { _all.total.docs.count: 2 } 
```
It was done to explicitly demonstrate that condition will be applied only after primaries reach max_docs.

> The rest test are run (sometimes) against a 1 node cluster. This means that replicas won't always be assigned. This wait_for_active_shards: all condition in the rest test makes it fail. I suggest just removing it. We also probably don't need explicit setting replicas to 1. Better leave to the default imo.

> we run test with one node and with multiple nodes I think your test will be fine it should just use defaults and remove the wait_for_active_shards: all condition to make sure it also passes in a single node enviroment

@s1monw @bleskes Thanks for suggestions!

**Comment by fred84** [comment]
":distribution:integ-test-zip:integTest" now works fine, but ":qa:mixed-cluster:integTest" fails most of the times locally. I will try to figure it out.

**Comment by fred84** [comment]
@bleskes I currently added "skip before 5.6.1 version" in yml to pass mixed cluster test, but I'm not sure that it is proper version.

**Review by bleskes — APPROVED** [review]

**Comment by bleskes** [comment]
@fred84 that makes sense for now, I'll fix it later based on how we decide to backport this. Thanks for all the itertations

**Pull request merged**